### PR TITLE
balancer: add ExitIdle optional interface

### PR DESCRIPTION
**Pull request by dfawley**
RELEASE NOTES:
- balancer: add ExitIdle interface to instruct the balancer to attempt to leave the IDLE state by connecting SubConns if appropriate.  The method will be required by the Balancer interface in the future.

### CODE REVIEW

File: clientconn.go [hunk 1]
```diff
@@ -555,14 +555,11 @@ func (cc *ClientConn) GetState() connectivity.State {
...
-	if cc.GetState() == connectivity.Idle {
-		cc.mu.Lock()
-		for ac := range cc.conns {
-			// TODO: should this be a signal to the LB policy instead?
-			go ac.connect()
-		}
-		cc.mu.Unlock()
+	cc.mu.Lock()
+	if cc.balancerWrapper != nil {
+		cc.balancerWrapper.updateCh.Put(exitIdle{})
```
Comment (menghanl) [inline clientconn.go:560]: Don't call the channel directly. ClientConn shouldn't know internal of balancerWrapper. Add a method to balancerWrapper.

And if that method returns a bool, you can keep `go ac.connect()` inside cc

```go
cc.mu.Lock()
defer cc.mu.Unlock()
if cc.GetState() != connectivity.Idle {
  return
}
if cc.balancerWrapper.exitIdle() { // This method will do type assertion
  return
}
// fallback
for ac := range ccb.cc.conns {
  go ac.connect()
}
```

File: xds/internal/balancer/cdsbalancer/cdsbalancer_test.go [hunk 2]
```diff
@@ -1,3 +1,4 @@
+//go:build go1.12
```
Comment (menghanl) [inline xds/internal/balancer/cdsbalancer/cdsbalancer_test.go:1]: Did you want to keep this?

File: xds/internal/balancer/cdsbalancer/cdsbalancer.go [hunk 3]
```diff
@@ -361,6 +363,14 @@ func (b *cdsBalancer) run() {
...
 				}
 				b.childLB.UpdateSubConnState(update.subConn, update.state)
+			case exitIdle:
+				if b.childLB == nil {
+					b.logger.Errorf("xds: received ExitIdle with no child balancer")
+					break
+				}
+				if ei, ok := b.childLB.(balancer.ExitIdle); ok {
+					ei.ExitIdle()
+				}
```
Comment (menghanl) [inline xds/internal/balancer/cdsbalancer/cdsbalancer.go:373]: Add a note that this assumes the child supports `ExitIdle()`.

File: xds/internal/balancer/cdsbalancer/cdsbalancer_test.go [hunk 4]
```diff
@@ -120,6 +121,10 @@ func (tb *testEDSBalancer) Close() {
 	tb.closeCh.Send(struct{}{})
 }
 
+func (tb *testEDSBalancer) ExitIdle() {
+	// TODO: test
```
Comment (menghanl) [inline xds/internal/balancer/cdsbalancer/cdsbalancer_test.go:125]: What's this TODO for?

File: xds/internal/balancer/clusterresolver/clusterresolver.go [hunk 5]
```diff
@@ -279,6 +281,14 @@ func (b *clusterResolverBalancer) run() {
...
 				}
 				b.child.UpdateSubConnState(update.subConn, update.state)
+			case exitIdle:
+				if b.child == nil {
+					b.logger.Errorf("xds: received ExitIdle with no child balancer")
+					break
+				}
+				if ei, ok := b.child.(balancer.ExitIdle); ok {
+					ei.ExitIdle()
+				}
```
Comment (menghanl) [inline xds/internal/balancer/clusterresolver/clusterresolver.go:291]: Add a note

File: xds/internal/balancer/clusterresolver/clusterresolver_test.go [hunk 6]
```diff
@@ -1,3 +1,4 @@
+//go:build go1.12
```
Comment (menghanl) [inline xds/internal/balancer/clusterresolver/clusterresolver_test.go:1]: And this?

**Review by menghanl — CHANGES_REQUESTED** [review]

File: clientconn.go (hunk 1 above)
Comment (dfawley) [reply clientconn.go:560]: Moved the updateCh pushing into ccBalancerWrapper, but I don't think things can change so drastically here (follow-up offline).

File: xds/internal/balancer/cdsbalancer/cdsbalancer.go (hunk 3 above)
Comment (dfawley) [reply xds/internal/balancer/cdsbalancer/cdsbalancer.go:373]: Done

File: xds/internal/balancer/clusterresolver/clusterresolver.go (hunk 5 above)
Comment (dfawley) [reply xds/internal/balancer/clusterresolver/clusterresolver.go:291]: Done

File: xds/internal/balancer/cdsbalancer/cdsbalancer_test.go (hunk 2 above)
Comment (dfawley) [reply xds/internal/balancer/cdsbalancer/cdsbalancer_test.go:1]: This is the new build syntax - gofmt on my machine adds it.  It's harmless and will be the new standard going forward so I think it's fine.

File: xds/internal/balancer/clusterresolver/clusterresolver_test.go (hunk 6 above)
Comment (dfawley) [reply xds/internal/balancer/clusterresolver/clusterresolver_test.go:1]: (And this.)

File: xds/internal/balancer/cdsbalancer/cdsbalancer_test.go (hunk 4 above)
Comment (dfawley) [reply xds/internal/balancer/cdsbalancer/cdsbalancer_test.go]: For me to add a test.  (Oops! Done)

File: balancer/balancer.go [hunk 7]
```diff
@@ -75,24 +75,29 @@ func Get(name string) Builder {
...
+// A SubConn represents a single connection to a gRPC backend service.
+//
+// Each SubConn contains a list of addresses.
+//
+// All SubConns start in IDLE, and will not try to connect. To trigger the
+// connecting, Balancers must call Connect.
+//
+// gRPC will try to connect to the addresses in sequence, and stop trying the
+// remainder once the first connection is successful. If an attempt to connect
+// to all addresses encounters an error, the SubConn will automatically attempt
```
Comment (menghanl) [inline balancer/balancer.go:87]: Will **not** automatically attempt?
It will transition to IDLE, right?

File: balancer/balancer.go [hunk 8]
```diff
@@ -326,6 +331,20 @@ type Balancer interface {
...
 }
 
+// ExitIdle is an optional interface for balancers to implement.  If
+// implemented, ExitIdle will be called when ClientConn.Connect is called, if
+// the ClientConn is idle.  If unimplemented, ClientConn.Connect will cause
+// all SubConns to connect.
+//
+// Notice: it will be required for all balancers to implement this in a future
+// release.
+type ExitIdle interface {
```
Comment (menghanl) [inline balancer/balancer.go:341]: `ExitIdle_r_`??? hehehe

File: balancer/balancer.go (hunk 7 above)
Comment (dfawley) [reply balancer/balancer.go]: Yes, I'm not sure what I was thinking.

File: balancer/balancer.go (hunk 8 above)
Comment (dfawley) [reply balancer/balancer.go]: Done

**Review by menghanl — APPROVED** [review]

**Pull request merged**